### PR TITLE
Add TypeScript guides

### PR DESCRIPTION
**Pull request by gitKrystan**
image

While we have [grand plans](https://github.com/orgs/emberjs/projects/9/views/1?pane=issue&itemId=37988245) to eventually have a js/ts toggle for all of the guides, we need _something_ to point to in the meantime. In this PR, I am adding initial TypeScript guides pulled from the various resources currently scattered across the interwebs to use as a starting point.

These pages currently will _not_ show up in the table of contents. You'll have to visit the link directly. The purpose of this is to allow us to merge this mega-PR and continue to iterate without being fully released. In the meantime, we can link people directly to these pages with the caveat that the pages are still under construction.

Sources:
* https://docs.ember-cli-typescript.com/
* https://blog.skylight.io/ts-extends-confidence-2-2022/

This will help us close out https://github.com/emberjs/rfcs/pull/952

## To review:
Visit deploy preview linked below and navigate to `./release/using-typescript/`
(e.g. https://deploy-preview-1960--ember-guides.netlify.app/release/using-typescript/ for the deploy that was current at the time of this writing)

Review for:
* Typos
* Mistakes
* Completeness

Please make liberal use of the GitHub suggestions feature because I am starting to get bleary-eyed looking at this for so long. 🙏

### CODE REVIEW

File: guides/release/using-typescript/configuration.md
Comment (chriskrycho) [inline guides/release/using-typescript/configuration.md:43]:
```suggestion
If you're migrating from `ember-cli-typescript`, particularly an older version, to Ember's out-of-the-box TypeScript support, you may also need to update your `tsconfig.json`. Current versions of `ember-cli-typescript` generate the correct config at installation. You do *not* need to install `ember-cli-typescript` for new apps or adding.
```

File: guides/release/using-typescript/getting-started.md
Comment (chriskrycho) [inline guides/release/using-typescript/getting-started.md:27]: This makes me extremely sad. 😂

File: guides/release/using-typescript/getting-started.md
Comment (chriskrycho) [inline guides/release/using-typescript/getting-started.md:13]:
```suggestion
- We configure Babel to transpile TypeScript files.
- We configure TypeScript to type check your project.
```

File: guides/release/using-typescript/getting-started.md
Comment (chriskrycho) [inline guides/release/using-typescript/getting-started.md:27]:
```suggestion

<div class="cta">
  <div class="cta-note">
    <div class="cta-note-body">
      <div class="cta-note-heading">Zoey says...</div>
      <div class="cta-note-message">
        Ember also includes its own types compiled directly from its source code, as described <a href="https://blog.emberjs.com/stable-typescript-types-in-ember-5-1/">in this blog post</a>. We still use the <code>@types</code> packages by default for the sake of compatibility with Ember Data, because Ember Data is not yet compatible with Ember’s native official types. However, if you only do not use Ember Data, we recommend following the instructions in that blog post to switch to the native types, which are guaranteed to always be 100% correct and 100% up to date!
      </div>
    </div>
    <img src="/images/mascots/zoey.png" role="presentation" alt="">
  </div>
</div>
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:4]:
```suggestion

**Note:** 🚧 This section is under construction! 🏗️ The content here may not be fully up to date!

```

File: guides/release/using-typescript/index.md
Comment (chriskrycho) [inline guides/release/using-typescript/index.md:1]:
```suggestion
**Note:** 🚧 This section is under construction! 🏗️ The content here may not be fully up to date!

This guide is designed to help you get up and running with TypeScript in an Ember app.

```

File: guides/release/using-typescript/getting-started.md
Comment (chriskrycho) [inline guides/release/using-typescript/getting-started.md:1]:
```suggestion
**Note:** 🚧 This section is under construction! 🏗️ The content here may not be fully up to date!

To start a new Ember project with TypeScript, you can add the `--typescript` flag when you run [`ember new`](../../getting-started/quick-start):
```

File: guides/release/using-typescript/configuration.md
Comment (chriskrycho) [inline guides/release/using-typescript/configuration.md:1]:
```suggestion
**Note:** 🚧 This section is under construction! 🏗️ The content here may not be fully up to date!

```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:1]:
```suggestion
**Note:** 🚧 This section is under construction! 🏗️ The content here may not be fully up to date!
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:70]:
```suggestion
You'll want to use other type definitions as much as possible. Many packages ship their own type definitions, and many others have community-maintained definitions from [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped), available in the `@types` name space. The first thing you should do is to look for types from other addons: it will mean writing `any` a lot less and getting a lot more help both from your editor and from the compiler.
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:72]:
```suggestion
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:73]:
```suggestion
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:102]:
```suggestion
### Classic `get` or `set` methods
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:104]:
```suggestion
In general, the `this.get` and `this.set` methods on `EmberObject` subclasses and the standalone `get` and `set` functions will work as you'd expect _if_ you're doing lookups only a single layer deep. We do not provide support for deep key lookups like `get(someObj, 'a.b.c')`, because normal property access can works correctly across the whole Ember ecosystem since at least Ember and Ember Data 3.28.
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:124]: Since regular property access “just works”, and has for a very long time, you should migrate to using normal property access instead. TypeScript will help make this a smooth process by identifying where you need to handle null and undefined intermediate properties.

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:130]:
```suggestion
The officially supported method for injections with TypeScript uses _decorators_ along with a “type registry” which maps the string names of services to the corresponding type. For example, a session service which defines a `login` method might be defined like this—notice the `declare module '@ember/service`:
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:157]:
```suggestion
import type MySession from 'my-app/services/my-session';
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:130]:
```suggestion
The officially supported method for injections with TypeScript uses _decorators_. Since decorators do not currently have access to enough information to produce an appropriate type, we need to import and name the type explicitly. For example, we might have `MySession` service which defines a `login` method, defined as usual:
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:173]:
~~~suggestion

Finally, you may have noticed the `declare module` at the bottom of the example `MySession` definition:

```typescript
declare module '@ember/service' {
  interface Registry {
    'my-session': MySession;
  }
}
```

This definition allows for type-safe lookups with other Ember dependency injection APIs. For example, [the `Owner.lookup` method](https://api.emberjs.com/ember/5.2/classes/Owner#2-method) uses this "registration"—a mapping from the string `'my-session'` to the service type, `MySession`—to provide the correct type:

```typescript
function dynamicLookup(owner: Owner) {
  let mySession = owner.lookup('service:my-session');
  mySession.login("tom@example.com", "password123");
}
```

~~~

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:193]: BALETED

```suggestion
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:128]:
```suggestion
Ember looks up services with the `@service` decorator at runtime, using the name of the service being injected up as the default value—a clever bit of metaprogramming that makes for a nice developer experience. TypeScript cannot do this, because the name of the service to inject isn't available at compile time in the same way. (These same considerations apply to controller injections using the `@inject` decorator from `@ember/controller`.)
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:196]:
```suggestion
We use the same basic approach for Ember Data type lookups with string keys as we do for service injections, but here we take advantage of the string "type registration" for the runtime code as well. As a result, once you add the module and interface definitions for each model, serializer, and adapter in your app, you will automatically get type-checking and autocompletion and the correct return types for functions like `findRecord`, `queryRecord`, `adapterFor`, `serializerFor`, etc. No need to try to write out those (admittedly kind of hairy!) types; just write your Ember Data calls like normal and everything _should_ just work.
```

File: guides/release/using-typescript/ts-and-ember.md
Comment (chriskrycho) [inline guides/release/using-typescript/ts-and-ember.md:317]: DELETE (long since irrelevant!)

```suggestion
```

**Review by chriskrycho — COMMENTED** [review]
Thanks for getting this started! I left a bunch of suggestions, and I think we want to keep these *out* of the top-level outline for now, but getting it started this way is awesome, and now we can iterate!

File: guides/release/using-typescript/signatures.md
Comment (chriskrycho) [inline guides/release/using-typescript/signatures.md:634]:
```suggestion
We can also define signatures in more complicated ways using more advanced TypeScript features.
Nearly anything you can do with a “regular” TypeScript function or class, you can also do with signatures for Glimmer invokables.
We can make a component accept a generic type, or use union types.
With these tools at our disposal, we can even define our signatures to [make illegal states unrepresentable][illegal].

To see this in practice, consider a list component which yields back out instances of the same type it provides, and provide the appropriate element target based on a `type` argument.
Yielding back out the same type passed in will use generics, and providing an appropriate element target for `...attributes` can use a union type.
```

File: guides/release/using-typescript/signatures.md
Comment (chriskrycho) [inline guides/release/using-typescript/signatures.md:704]:
```suggestion

The same approach with generics works for class-based helpers and class-based modifiers.
Function-based helpers as and modifiers can also use generics, but by using them on the function definition rather than via a signature.
One caveat: particularly complicated union types in signatures can sometimes become too complex for Glint/TypeScript to resolve when invoking in a template.
In those cases, your best bet is to find a simpler way to structure the types while preserving safety.

```

File: guides/release/using-typescript/signatures.md
Comment (chriskrycho) [inline guides/release/using-typescript/signatures.md]: Honestly, the "First… Second… Third" is probably better anyway!

File: guides/release/using-typescript/signatures.md
Comment (chriskrycho) [inline guides/release/using-typescript/signatures.md]: Grammar nit:

```suggestion
Let’s go one step further and switch to supporting for two [named blocks][nb]: an optional `title` block for a caption for the audio element, and a `fallback` block for the audio fallback where we previously used a `default` block.
```

File: guides/release/using-typescript/core-concepts/routing.md
Comment (chriskrycho) [inline guides/release/using-typescript/core-concepts/routing.md:83]: We should update this to use the `Awaited<T>` type that TS ships natively as of late in the 4.X series!

**Comment by jenweber** [comment]
Wow, this is amazing and there is so much content here! Incredible work.

What kind of feedback do you want the most at this stage? When do you think this might be ready to merge? (Now, after more feedback, do you know of any blockers, etc)

**Comment by gitKrystan** [comment]
IMO this is ready to merge, but no one else has read it.

**Review by MinThaMie — COMMENTED** [review]
Discussed today during the learning meeting
- Remove the "under construction message"
- Add it to the table of content
  - Don't call them the same as in the current menu, but make it something like "TypeScript & EmberData"
 
 We would love to merge it sooner rather then later, because it looks great!

**Comment by gitKrystan** [comment]
Thanks! Done!

image

File: guides/release/using-typescript/additional-resources/legacy.md
Comment (gitKrystan) [inline guides/release/using-typescript/additional-resources/legacy.md:197]:
```suggestion
[migrating-from-classic-classes]: ../../../upgrading/current-edition/native-classes/
```

File: guides/release/pages.yml
Comment (mansona) [inline guides/release/pages.yml:253]: one thing that is important once we release something is the URLs, we need to always keep redirects and never break a link. This means that we need to be sure that we are happy with the URLs that we are generating.

Personally I think `using` in this URL is too much 🤔 i.e. `/release/using-typescript/` is not as good as `/release/typescript/`. What do you think?

File: guides/release/using-typescript/application-development/index.md
Comment (mansona) [inline guides/release/using-typescript/application-development/index.md:2]: same here, why do we have this redirect?

File: guides/release/using-typescript/additional-resources/index.md
Comment (mansona) [inline guides/release/using-typescript/additional-resources/index.md:2]: why do we have a redirect for a page that hasn't launched yet? What I said before about "never breaking a link" is true, but we don't need to do that for things that are not released yet 🤔

File: guides/release/using-typescript/application-development/converting-an-app.md
Comment (gitKrystan) [inline guides/release/using-typescript/application-development/converting-an-app.md:7]: This approach kinda stinks but I experimented w/ different ways and this was the least obnoxious one. I have it on my TODO list to add an appropriate flag...probably to ember-cli-update.

**Review by mansona — COMMENTED** [review]
Just some minor comments, I haven't reviewed the contents (I'm not sure I'm qualified 😂)

One other thing it might be worth considering is to split up the invokables docs into 2 pages because the diff was so big that github didn't want to show it at first 🙈

File: guides/release/using-typescript/application-development/index.md
Comment (gitKrystan) [reply guides/release/using-typescript/application-development/index.md:2]: IIRC clicking the top-level header was broken w/o an "index.md". Will double-check.

Comment (mansona) [reply guides/release/using-typescript/application-development/index.md:2]: ah yes that will probably be true, but there should be an index and it should have content instead of a redirect ;) in this case you would just need to rename the `configuration` file to be `index.md` and then that would be the first thing in the section 👍 the name of the page can still be `Configuration` it just changes the url structure

Comment (gitKrystan) [reply guides/release/using-typescript/application-development/index.md:2]: I have some ideas to fix. Please hold.

File: guides/release/typescript/additional-resources/legacy.md
Comment (gitKrystan) [inline guides/release/typescript/additional-resources/legacy.md:166]:
```suggestion
The [`this` type][this], tells TS to use `UserProfile` for `get` and `set` lookups; otherwise `this.get` would not know the types of `'name'` or `'age'` or even be able to suggest them for autocompletion.
```

File: guides/release/typescript/application-development/addons.md
Comment (gitKrystan) [inline guides/release/typescript/application-development/addons.md:173]:
```suggestion
[this]: https://www.typescriptlang.org/docs/handbook/2/functions.html#declaring-this-in-a-function
[typesVersions]: https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#version-selection-with-typesversions
```

File: guides/release/typescript/core-concepts/ember-data.md
Comment (runspired) [inline guides/release/typescript/core-concepts/ember-data.md:31]:
```suggestion
  @attr
```

File: guides/release/typescript/core-concepts/ember-data.md
Comment (lukasnys) [inline guides/release/typescript/core-concepts/ember-data.md:63]: Following runspired's suggestion for removing parentheses above, should they also be removed here?

```suggestion
  @attr
```

File: guides/release/typescript/core-concepts/ember-data.md
Comment (lukasnys) [inline guides/release/typescript/core-concepts/ember-data.md:213]:
```suggestion
EmberData serializers and adapters are normal TypeScript classes. The only related gotcha is that you must [register][ED-registry] them with a declaration:
```

File: guides/release/typescript/core-concepts/invokables.md
Comment (lukasnys) [inline guides/release/typescript/core-concepts/invokables.md:711]:
```suggestion
Function-based helpers and modifiers can also use generics, but by using them on the function definition rather than via a signature.
```

**Review by lukasnys — CHANGES_REQUESTED** [review]

**Comment by lukasnys** [comment]
Very excited to see this land! 🚀

File: guides/release/typescript/index.md
Comment (jaredgalanis) [inline guides/release/typescript/index.md:3]: This makes it sound like we are linking to the Ember Guides as a separate resource, but I think we're actually making this a part of the Ember Guides. Maybe:

```suggestion
This is _not_ an introduction to TypeScript _or_ Ember. Throughout this guide, we’ll link back to [the TypeScript docs][typescript-docs] and to other sections of [the Ember Guides][ember-guides] when there are specific concepts that we will not explain here but which are important for understanding what we’re covering!
```

File: guides/release/typescript/core-concepts/index.md
Comment (jaredgalanis) [inline guides/release/typescript/core-concepts/index.md:3]: Same, concern here as https://github.com/ember-learn/guides-source/pull/1960/files#r1367496976. 

Maybe:

```suggestion
We do _not_ cover general usage of Ember; instead, we assume that as background knowledge. Please see other sections of the [Ember Guides][ember-guides] and [API docs][api-docs]!
```

File: guides/release/typescript/core-concepts/services.md
Comment (jaredgalanis) [inline guides/release/typescript/core-concepts/services.md:5]: Same concern. Maybe:

```suggestion
Let's take this example from elsewhere in the [Ember Guides][example-location]:
```

**Review by jaredgalanis — COMMENTED** [review]
@gitKrystan, I've made a few comments that I wouldn't classify as blocking. These guides are incredible, thank you so much for doing this! 🔥🎉

File: guides/release/typescript/index.md
Comment (chriskrycho) [reply guides/release/typescript/index.md:3]: Heh, that's because this content is all originally pulled from (though *massively* updated, thanks to @gitKrystan’s amazing efforts) https://docs.ember-cli-typescript.com (which can go away/redirect here once this lands!). So originally it *was* just pointing the the Ember Guides as a separate resource. 😅

**Comment by gitKrystan** [comment]
Lint should be fixed now.

I played w/ moving the chapter up a level (per discussion on Discord) and it makes the initial TOC super long + pushes some of the TypeScript chapters to the same level as their top-level equivalent (e.g. "Getting Started" at the same level as "TypeScript: Getting Started") which I found a bit confusing.

**Pull request merged**
